js/src/cli-util/pull.ts

```diff
@@ -0,0 +1,303 @@
+import {
+  FunctionObject,
+  functionSchema,
+  SavedFunctionId,
+} from "@braintrust/core/typespecs";
+import { _internalGetGlobalState } from "../logger";
+import { loadCLIEnv } from "./bundle";
+import { PullArgs } from "./types";
+import { warning } from "../framework";
+import { z } from "zod";
+import { ProjectNameIdMap } from "../functions/upload";
+import fs from "fs";
+import util from "util";
+import slugify from "slugify";
+import path from "path";
+import { currentRepo } from "../gitutil";
+import { isEmpty } from "@braintrust/core";
+import {
+  ToolFunctionDefinition,
+  toolFunctionDefinitionSchema,
+} from "../framework2";
+
+export async function pullCommand(args: PullArgs) {
+  await loadCLIEnv(args);
+
+  const loggerConn = _internalGetGlobalState().apiConn();
+  const functions = await loggerConn.get_json("/v1/function", {
+    ...(args.project_id ? { project_id: args.project_id } : {}),
+    ...(args.project_name ? { project_name: args.project_name } : {}),
+    ...(args.slug ? { slug: args.slug } : {}),
+    ...(args.id ? { ids: [args.id] } : {}),
+  });
+  const functionObjects = z
+    .object({ objects: z.array(z.unknown()) })
+    .parse(functions);
+
+  const projectNameToFunctions: Record<string, FunctionObject[]> = {};
+  const projectNameIdMap = new ProjectNameIdMap();
+
+  for (const rawFunc of functionObjects.objects) {
+    const parsedFunc = functionSchema.safeParse(rawFunc);
+    if (!parsedFunc.success) {
+      const id =
+        typeof rawFunc === "object" && rawFunc && "id" in rawFunc
+          ? ` ${rawFunc.id}`
+          : "";
+      console.warn(
+        warning(`Failed to parse function${id}: ${parsedFunc.error.message}`),
+      );
+      continue;
+    }
+
+    const func = parsedFunc.data;
+    const projectName = await projectNameIdMap.getName(func.project_id);
+    if (!projectNameToFunctions[projectName]) {
+      projectNameToFunctions[projectName] = [];
+    }
+    projectNameToFunctions[projectName].push(func);
+  }
+
+  console.log("Found functions in the following projects:");
+  for (const projectName of Object.keys(projectNameToFunctions)) {
+    console.log(` * ${projectName}`);
+  }
+
+  const outputDir = args.output_dir ?? "./braintrust";
+  fs.mkdirSync(outputDir, { recursive: true });
+
+  const git = await currentRepo();
+  const diffSummary = await git?.diffSummary("HEAD");
+
+  // Get the root directory of the Git repository
+  const repoRoot = await git?.revparse(["--show-toplevel"]);
+
+  const dirtyFiles = new Set(
+    (diffSummary?.files ?? []).map((f) =>
+      path.resolve(repoRoot ?? ".", f.file),
+    ),
+  );
+
+  for (const projectName of Object.keys(projectNameToFunctions)) {
+    const projectFile = path.join(
+      outputDir,
+      `${slugify(projectName, { lower: true, strict: true, trim: true })}.ts`,
+    );
+    const resolvedProjectFile = path.resolve(projectFile);
+    const fileExists = fs.existsSync(projectFile);
```

Comment:
Why use the synchronous fs api in an async function?

Reply:
sure fixed